framework/include/utils/SwapBackSentinel.h

```diff
@@ -0,0 +1,72 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#ifndef SWAPBACKSENTINEL_H
+#define SWAPBACKSENTINEL_H
+
+// MOOSE includes
+#include "FEProblem.h"
+
+/**
+ * The "SwapBackSentinel" class's destructor guarantees that
+ * FEProblem::swapBackMaterials{Face,Neighbor}() is called even when
+ * an exception is thrown from FEProblem::reinitMaterials{Face,Neighbor}.
+ * This is because stack unwinding (for caught exceptions) guarantees
+ * that object destructors are called.  The typical way of using this
+ * object is to construct it in the same scope where reinitMaterials
+ * is called:
+ *
+ * {
+ *   SwapBackSentinel sentinel(_fe_problem, &FEProblem::swapBackMaterials, _tid);
+ *   _fe_problem.reinitMaterials(_subdomain, _tid);
+ * }
+ */
+class SwapBackSentinel
+{
+public:
+  /**
+   * SwapBackFunction is a typedef for a pointer to an FEProblem
+   * member function taking a THREAD_ID and returning void.  All the
+   * FEProblem::swapBackMaterialXXX() members have this signature.
+   */
+  typedef void (FEProblem::*SwapBackFunction)(THREAD_ID);
```

Comment:
I looked into it, and the syntax was [way worse](https://github.com/jwpeterson/moose/commit/1a5b692bc839c114e5d476eb7b9f5767dc20ebdd) for dealing with member function pointers IMO.  I'd be happy to change it if you can figure out an interface that is cleaner than the C++03 way for this simple application...